Enable Error Prone check: UnnecessaryStringBuilder

ksobolew opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

This is a new check in Error Prone 2.20.0. Per description: Prefer string concatenation over explicitly using `StringBuilder#append`, since `+` reads better and has equivalent or better performance.

Ever since Java 9 there are specialized optimizations for String concatenation and it no longer desugars to `StringBuilder` usage. This commit fixes mostly some very old-fashioned code where `StringBuilder` never really made sense anyway.

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues


<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

findepi commented on core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java:
```diff
@@ -727,17 +726,15 @@ private String visitFilter(Expression node, Void context)
-                    .append(process(node.getSubquery(), context))
-                    .append(")")
-                    .toString();
+            return "(" +
+                    process(node.getValue(), context) +
+                    ' ' +
+                    node.getOperator().getValue() +
+                    ' ' +
+                    node.getQuantifier().toString() +
+                    ' ' +
+                    process(node.getSubquery(), context) +
+                    ")";
```
would this read better with `"(%s %s %s %s)".formatted(....)`?

ksobolew replied:
That's what I suspected. But that means no `String#formatted` (added in Java 15). `String#format` it is!

electrum commented on plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:
```diff
@@ -86,21 +86,16 @@ public static String generatePql(PinotTableHandle tableHandle, List<PinotColumnH
-        return tableHandle.getTableName();
+        return tableNameSuffix.map(suffix -> new StringBuilder(tableHandle.getTableName())
+                .append(suffix)
+                .toString())
```
No need to use `StringBuilder` here:
```java
return tableNameSuffix
        .map(suffix -> tableHandle.getTableName() + suffix)
        .orElseGet(tableHandle::getTableName);
```
Or we could do
```java
return tableHandle.getTableName() + tableNameSuffix.orElse("");

ksobolew replied:
> No need to use StringBuilder here:

This commit is just an preparatory step before converting to exactly this in the next commit. I didn't want to conflate these two in a single step.

electrum commented on plugin/trino-kinesis/src/main/java/io/trino/plugin/kinesis/KinesisShardCheckpointer.java:
```diff
@@ -80,14 +80,13 @@ public KinesisShardCheckpointer(
-                .append("_")
-                .append(String.valueOf(iterationNo))
-                .toString();
+        return this.logicalProcessName +
```
```java
return "%s_%s_%s_%s".formatted(logicalProcessName, kinesisSplit.getStreamName(), kinesisSplit.getShardId(), iterationNo);

ksobolew commented:
AC

martint commented on core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java:
```diff
@@ -727,17 +726,12 @@ private String visitFilter(Expression node, Void context)
-                    .append(process(node.getSubquery(), context))
-                    .append(")")
-                    .toString();
+            return format(
+                    "(%s %s %s %s)",
+                    process(node.getValue(), context),
+                    node.getOperator().getValue(),
+                    node.getQuantifier(),
+                    process(node.getSubquery(), context));
```
`"...".formatted(...)`

ksobolew replied:
`String#formatted` was introduced in Java 13, but `trino-parser` is deliberately at Java 8 language level.

ksobolew replied:
After `trino-grammar` was extracted from `trino-parser`, this is now possible.

ksobolew commented:
Resolved conflicts

mosabua commented:
@ksobolew is this still in progress and valid?

findinpath commented:
@ksobolew pls rebase on top of `master`

ksobolew commented:
There you go

findinpath commented on core/trino-spi/src/main/java/io/trino/spi/procedure/Procedure.java:
```diff
@@ -99,14 +99,12 @@ public boolean requiresNamedArguments()
-                .append(arguments.stream()
+        return schema + '.' + name +
+                '(' +
+                arguments.stream()
```
why not stringjoiner?

ksobolew replied:
It was just a mechanical transformation. If I was taking it further I would rather use `Collectors#joining(delim, prefix, suffix)` instead.

findinpath commented:
@pettyjamesm you've had a few contributions related to migrating towards string concatention. Would you please take a look over this code cosmetics PR and accompany it over the finish line? :pray:

pettyjamesm commented on core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:
```diff
@@ -549,11 +549,9 @@ public Map<SchemaTableName, RelationType> getRelationTypes(Session session, Qual
-            return ImmutableMap.of();
+            return getRelationTypeIfExists(session, objectName.get())
+                    .map(type -> ImmutableMap.of(objectName.get().asSchemaTableName(), type))
+                    .orElseGet(ImmutableMap::of);
```
`ImmutableMap.of()` is a singleton (ie: has no allocation overhead), so `.orElse(ImmutableMap.of())` is cheaper than `.orElseGet(ImmutableMap::of)`

pettyjamesm commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/rest/TrinoRestCatalog.java:
```diff
@@ -549,12 +549,11 @@ private static TableIdentifier toIdentifier(SchemaTableName schemaTableName)
-        }
-
-        return ImmutableList.of(Namespace.of(namespace.get()));
+        return namespace
```
This seems significantly harder to read to me compared to the previous version with the conditional

ksobolew replied:
It's a matter of taste, really. I prefer the functional style, personally. I reverted all changes from this commit except in the file where it's needed for further changes (they were also requested by a maintainer).

pettyjamesm commented on plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:
```diff
@@ -86,21 +86,15 @@ public static String generatePql(PinotTableHandle tableHandle, List<PinotColumnH
-        return tableHandle.getTableName();
+        return tableNameSuffix
+                .map(suffix -> new StringBuilder(tableHandle.getTableName()).append(suffix).toString())
+                .orElseGet(tableHandle::getTableName);
```
Same thing here about the conditional style seeming much easier to read and understand than the functional style

ksobolew commented:
> I'm also not sure I understand the motivation to switch from `format("", ...)` to `"".formatted(...)` since the two seem to be equivalent.

The sequence of events was like this:

* I converted one of the `StringBuilder`s to `String#format` (the old style), like everything else in this module
* A maintainer said "Let's make it `String#formatted`"
* I converted the module to `String#formatted` for consistency

Note that in the meantime the module's Java language level was bumped from 8 to 17.

mosabua commented:
Are you going to merge this @pettyjamesm or do you want any others to chime in?

pettyjamesm commented:
Merged, thanks @ksobolew !
